Add flush memory threshold for appender to prevent OOM

p1p1bear opened the pull request:
Currently, memory in the appender is only flushed after every 204,800 appends, and FlushChunk is triggered every 2,048 appends. This delayed flushing behavior can lead to excessive memory consumption and increase the risk of out-of-memory in high-throughput ingestion scenarios.

This patch introduces a new configuration option for appender, which sets a peak memory threshold. When the appender's memory usage exceeds this limit, it forces an early flush of temporary allocations — even within a single chunk.

Additionally, this threshold now influences the early triggering of FlushChunk, allowing more responsive memory management based on actual usage rather than fixed batch counts.

This change improves memory predictability and helps prevent unbounded growth during large appends.

Mytherin commented:
Thanks for the PR!

Could you make this a setting within the appender itself, rather than a DuckDB setting?

p1p1bear commented:
> Thanks for the PR!
> 
> Could you make this a setting within the appender itself, rather than a DuckDB setting?

Hi, Mytherin.
Thank you for the feedback.
Done — flush_memory_threshold is now a constructor parameter instead of a global option, as suggested.

p1p1bear commented:
Fix the format.

Mytherin reviewed (commented):
Thanks for the changes - I think this is almost there. Some more comments:

p1p1bear committed:
Add flush memory threshold for appender to prevent OOM

Currently, memory in the appender is only flushed after every 204,800 appends,
and FlushChunk is triggered every 2,048 appends. This delayed flushing behavior
can lead to excessive memory consumption and increase the risk of out-of-memory
in high-throughput ingestion scenarios.

This patch introduces a new configuration option for appender, which sets a
peak memory threshold. When the appender's memory usage exceeds this limit,
it forces an early flush of temporary allocations — even within a single chunk.

Additionally, this threshold now influences the early triggering of FlushChunk,
allowing more responsive memory management based on actual usage rather than
fixed batch counts.

This change improves memory predictability and helps prevent unbounded growth
during large appends.

p1p1bear commented:
Hi @Mytherin 

I've checked my [local CI](https://github.com/p1p1bear/duckdb/actions) and all tests have passed successfully. From the failed checks in the triggered pipeline, it seems the failure might not be related to the changes in this PR.

<img width="1085" height="106" alt="image" src="https://github.com/user-attachments/assets/afeba3f2-ce0b-4243-be31-16d317acf4ec" />

Mytherin commented:
Thanks!
